DDF-1886 Added admin ui for adding/editing local registry node information

clockard opened the pull request:
#### What does this PR do?
Adds admin ui for adding/editing local registry node information. 
#### Who is reviewing it (please choose AT LEAST two reviewers that need to approve the PR before it can get merged; if a component team is listed, at least one of its members needs to approve)?
@djblue @andrewkfiedler @harrison-tarr 
#### Choose 2 committers to review/merge the PR (please choose ONLY two committers from below, delete the rest).
@pklinef
@shaundmorris

#### How should this be tested?
Start up ddf and then start the DDF Registry app. Once the app is started you can open up the app and see a new 'Local Node Information' tab. You can play around with setting/adding/removing nodes/fields and saving them.
#### Any background context you want to provide?
This was pulled from the ddf-registry repo.
#### What are the relevant tickets?
DDF-1886
#### Screenshots (if appropriate)
Main table view
![screen shot 2016-06-06 at 4 08 54 pm](https://cloud.githubusercontent.com/assets/5248090/15840775/0fd42654-2c01-11e6-9587-a961fdc7a014.png)

Edit dialog
![screen shot 2016-06-06 at 4 01 43 pm](https://cloud.githubusercontent.com/assets/5248090/15840776/14d93446-2c01-11e6-88ba-53e6630d65f3.png)

Validation error
![screen shot 2016-06-07 at 4 11 57 pm](https://cloud.githubusercontent.com/assets/5248090/15882626/472dcb2e-2cf5-11e6-954a-f7f1ff5a1361.png)


#### Checklist:
- [ ] Documentation Updated
- [ ] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

vinamartin commented on catalog/spatial/registry/registry-admin-modules/registry-admin-local-ui/src/main/webapp/js/view/Registry.view.js:
```diff
@@ -0,0 +1,200 @@
+            editNode: function(evt) {
+                evt.stopPropagation();
+                var node = this.model;
+                //node.refreshData();
```
Get rid of commented out code

clockard replied:
Removed

vinamartin commented on catalog/spatial/registry/registry-admin-modules/registry-admin-local-ui/src/main/webapp/templates/customFieldList.handlebars:
```diff
@@ -0,0 +1,40 @@
+            <option value="number">number</option>
+            <option value="date">date</option>
+            <option value="point">point</option>
+            <!--<option value="bounds">bounds</option>-->
```
Do we need this? If not, delete

clockard replied:
Nope, removed.

vinamartin commented on catalog/spatial/registry/registry-admin-modules/registry-admin-local-ui/src/main/webapp/templates/nodeList.handlebars:
```diff
@@ -0,0 +1,30 @@
+    <th class="name">Name</th>
+    <th class="created">Created On</th>
+    <th class="modified">Last Modified</th>
+    <th><!--<a href="#" class="button-icon uploadNodeLink fa fa-upload"></a>-->
```
Do we need this?

clockard replied:
Nope, removed

shaundmorris commented on catalog/spatial/registry/registry-admin-modules/registry-admin-local-ui/pom.xml:
```diff
@@ -0,0 +1,132 @@
+                <configuration>
+                    <instructions>
+                        <!--we will use this servlet filter once it has been made generic to protect this app-->
+                        <!--<Embed-Dependency>search-securityservletfilter</Embed-Dependency>-->
```
Is this comment accurate?

clockard replied:
A copy past issue. Removed.

shaundmorris commented on catalog/spatial/registry/registry-admin-modules/registry-admin-local-ui/README.md:
```diff
@@ -0,0 +1,15 @@
+<!--
```
What is the plan for this README.md? Seems like it may not be needed.

clockard replied:
This was copied over from another module to be consistent

shaundmorris commented on catalog/spatial/registry/registry-admin-modules/registry-admin-local-ui/server-impl.js:
```diff
@@ -0,0 +1,115 @@
+server.requestProxy = function (req, res) {
+    "use strict";
+
+    req.url = "https://localhost:8993" + req.url;
```
Is this URL hardcoded? Will it be an issue if the hostname or port number changes?

clockard replied:
This is for running grunt. This file was copied from another module and I am pretty sure they all do this.

mcalcote commented on catalog/spatial/registry/registry-admin-modules/registry-admin-local-ui/src/main/java/org/codice/ddf/registry/admin/plugin/RegistryPlugin.java:
```diff
@@ -0,0 +1,36 @@
+     * Constructor.
+     */
+    public RegistryPlugin() {
+        this.displayName = "Local Registry Nodes";
```
Preemptively change this to "Local Node Information" based on feedback?

clockard replied:
Fixed

mcalcote commented on catalog/spatial/registry/registry-admin-modules/registry-admin-local-ui/src/main/webapp/WEB-INF/web.xml:
```diff
@@ -0,0 +1,27 @@
+         version="2.4">
+
+    <description>Admin UI - Local Registry Nodes</description>
+    <display-name>Admin UI - Local Registry Nodes</display-name>
```
Preemptively change this to "Local Node Information" based on feedback?

clockard replied:
Fixed

shaundmorris commented:
@clockard you'll have to resolve the merge conflicts on this PR at some point.

clockard commented:
@shaundmorris I was going to wait until it we can actually build and test it before I worry about merge conflicts since they are most likely in the pom with all the new registry stuff coming in.

djblue commented on catalog/spatial/registry/registry-admin-modules/registry-admin-local-ui/src/main/webapp/js/model/Association.js:
```diff
@@ -0,0 +1,251 @@
+    'js/model/FieldDescriptors.js',
+    'js/model/Segment.js',
+    'wreqr',
+    'jquery',
```
Looks like this dependency isn't being used in this module. Could remove.

clockard replied:
Removed

djblue commented on catalog/spatial/registry/registry-admin-modules/registry-admin-local-ui/src/main/webapp/js/model/FieldDescriptors.js:
```diff
@@ -0,0 +1,409 @@
+        },
+        constructEmailTitle: function () {
+            var title = this.getField('type').get('value') + ' ' + this.getField('address').get('value');
+            title = title.replace(/undefined/g, '').trim();
```
I think we don't have to remove undefined from the `title`, if we guard against `undefined` values.

Something like `this.getField('type').get('value') || ''` should do the trick.

clockard replied:
I was just trying to make it simpler by not checking every value i put into these title strings and just removing any 'undefined' values afterward. I used this technique in several methods in this file so they should all be consistent on how they do this.

clockard replied:
Fixed this by using an array to build up the title and then filtering out any undefined values like you recommended.

djblue commented on catalog/spatial/registry/registry-admin-modules/registry-admin-local-ui/src/main/webapp/js/view/Association.view.js:
```diff
@@ -0,0 +1,90 @@
+                this.simpleId = this.parentId.split(':').join('-');
+                this.listenTo(wreqr.vent, 'removeAssociation:' + this.parentId, this.removeAssociation);
+                this.listenTo(wreqr.vent, 'associationSegmentRemoved', this.updateAssociations);
+                this.listenTo(wreqr.vent, 'associationSegmentAdded', this.updateAssociations);
```
Could also:

```javascript
this.listenTo(this.collection, 'add', this.render);
this.listenTo(this.collection, 'remove', this.render);
```

to re-render whenever the collection changes.

clockard replied:
Added

djblue commented on catalog/spatial/registry/registry-admin-modules/registry-admin-local-ui/src/main/webapp/js/view/Association.view.js:
```diff
@@ -0,0 +1,90 @@
+            },
+            removeAssociation: function (associationId) {
+                this.collection.remove(this.model.removeAssociation(associationId));
+                this.render();
```
Not required, if the collection is listened to.

clockard replied:
Removed

djblue commented on catalog/spatial/registry/registry-admin-modules/registry-admin-local-ui/src/main/webapp/js/view/CustomizableField.view.js:
```diff
@@ -0,0 +1,101 @@
+            onRender: function(){
+                var customFields = [];
+                var allFields = this.model.get('fields').models;
+                _.each(allFields, function (field) {
```
This `_.each` feels like a filter. Something like:

```javascript
var customFields = this.model.get('fields')
  .filter(function (field) {
    return field.get('custom') !== undefined;
  });
```

might be more idiomatic.

clockard replied:
I like it. Changed

djblue commented on catalog/spatial/registry/registry-admin-modules/registry-admin-local-ui/src/main/webapp/js/view/Registry.view.js:
```diff
@@ -0,0 +1,199 @@
+                var view = this;
+                view.secondaryNodes = [];
+                var identityNode;
+                _.each(view.model.models, function(model){
```
Can this logic be moved out of `onRender`? It looks like other methods depend on `secondaryNodes`, which is being initialized in onRender.

clockard replied:
Moved this logic out to two separate functions getSecondaryNodes and getIdentityNode

djblue commented on catalog/spatial/registry/registry-admin-modules/registry-admin-local-ui/src/main/webapp/js/view/Segment.view.js:
```diff
@@ -0,0 +1,183 @@
+             */
+            setupPopOvers: function () {
+                var view = this;
+                this.model.get('fields').models.forEach(function (each) {
```
In `serializeData`, `this.model` is being checked, why is it not being checked here? Should it not be checked in `serializeData`?

clockard replied:
Not sure why we have the check in serializeData. I removed it. Changed the models.forEach to each since it is a backbone collection

stustison commented on catalog/spatial/registry/registry-admin-modules/registry-admin-local-ui/bower.json:
```diff
@@ -0,0 +1,37 @@
+  "version": "0.0.1",
+  "private": true,
+  "dependencies": {
+    "backbone-relational": "0.8.8",
```
You don't need both relational and associations. Which one are you actually using?

clockard replied:
Removed relational since we are using associations

stustison commented on catalog/spatial/registry/registry-admin-modules/registry-admin-local-ui/bower.json:
```diff
@@ -0,0 +1,37 @@
+    "require-css": "0.1.5",
+    "spin.js": "1.3.3",
+    "bootswatch": "=v3.1.1",
+    "jquery-file-upload": "9.5.7",
```
Are you actually uploading files?

clockard replied:
Initially we were going to but it didn't make it into the ui yet. I will remove it for now.

stustison commented on catalog/spatial/registry/registry-admin-modules/registry-admin-local-ui/src/main/webapp/js/model/Association.js:
```diff
@@ -0,0 +1,251 @@
+        defaults: {
+            associations: [],
+            associationSegments: [],
+            topSegment: null
```
Why null and not undefined?

clockard replied:
I thought that was how you defined defaults when it was an associated model. I got that from http://dhruvaray.github.io/backbone-associations/specify-associations.html. Should I change it to undefined?

stustison replied:
Just seemed weird to me. I want to say that I've always used undefined.

stustison commented on catalog/spatial/registry/registry-admin-modules/registry-admin-local-ui/src/main/webapp/js/model/Node.js:
```diff
@@ -0,0 +1,273 @@
+        },
+        sync: function () {
+
+            var deferred = $.Deferred();
```
You don't need this deferred.

clockard replied:
For some reason I wasn't returning the deferred like i was suppose to for the sync function. I am leaving it in now that it is the methods return value.

stustison commented on catalog/spatial/registry/registry-admin-modules/registry-admin-local-ui/src/main/webapp/js/model/Node.js:
```diff
@@ -0,0 +1,273 @@
+            return raw.value.localNodes;
+        },
+        deleteNodes: function (nodes) {
+            var deferred = $.Deferred();
```
Remove this deferred.

clockard replied:
Same for this deferred. I am now returning it.

stustison commented on catalog/spatial/registry/registry-admin-modules/registry-admin-local-ui/src/main/webapp/js/model/Node.js:
```diff
@@ -0,0 +1,273 @@
+                data: data,
+                url: this.deleteUrl
+            }).done(function (result) {
+                wreqr.vent.trigger('nodeDeleted');
```
You probably want this to trigger on success as opposed to just done.

clockard replied:
Changed to success

stustison commented on catalog/spatial/registry/registry-admin-modules/registry-admin-local-ui/src/main/webapp/js/model/Segment.js:
```diff
@@ -0,0 +1,464 @@
+                addSlotFields(fieldList, dataModel, descriptors[segType], model.get('segmentId'));
+
+                model.set('segments', segs);
+                model.set('fields', fieldList);
```
Every time you call set, you trigger a change event and subsequently a render, if something is hooked up to this model. You should be batching all of the model.set calls together and only calling set once with all of the changes.

clockard replied:
batched up sets into one set

stustison commented on catalog/spatial/registry/registry-admin-modules/registry-admin-local-ui/src/main/webapp/js/model/Segment.js:
```diff
@@ -0,0 +1,464 @@
+            var segs = [];
+            var seg;
+            var segType = model.get('segmentType');
+            model.dataModel = dataModel;
```
Something doesn't seem quite right with this. What is this dataModel variable that is attached to the model? Generally speaking, if you're attaching something directly to the backbone model itself, you're doing it wrong.

clockard replied:
Changed this to 'backingData' and added it to the model itself.
Same with the descriptors.

stustison commented on catalog/spatial/registry/registry-admin-modules/registry-admin-local-ui/src/main/webapp/js/view/NodeModal.view.js:
```diff
@@ -0,0 +1,141 @@
+                if (this.model) {
+                    data = this.model.toJSON();
+                }
+                data.saveErrors = this.model.saveErrors;
```
Again, we shouldn't ever be attaching things directly to the model itself.

clockard replied:
This should have been a NodModal.View variable. Changed it to be one.

stustison commented on catalog/spatial/registry/registry-admin-modules/registry-admin-local-ui/src/main/webapp/js/view/Segment.view.js:
```diff
@@ -0,0 +1,183 @@
+                if (this.model) {
+                    data = this.model.toJSON();
+                }
+                data.editableSegment = this.editableSegment;
```
This isn't the correct way to do this. Why not just have editableSegment in the model have the data that you want?

clockard replied:
After reviewing the code, found that this field actually mirrored one that was in the model already so just removed it all together.

stustison commented on catalog/spatial/registry/registry-admin-modules/registry-admin-local-ui/src/main/webapp/js/view/Segment.view.js:
```diff
@@ -0,0 +1,183 @@
+            },
+            removeSegment: function (id) {
+                this.model.removeSegment(id);
+                this.render();
```
You could just listen to the model and the view will call render when the field you care about changes.

clockard replied:
Changed to listen to the model and removed the render call.

stustison commented on catalog/spatial/registry/registry-admin-modules/registry-admin-local-ui/src/main/webapp/js/view/Segment.view.js:
```diff
@@ -0,0 +1,183 @@
+                this.model.removeSegment(id);
+                this.render();
+            },
+            serializeData: function () {
```
Why are you serializing data in a view? that is a model concern.

stustison commented on catalog/spatial/registry/registry-admin-modules/registry-admin-local-ui/src/main/webapp/main.js:
```diff
@@ -0,0 +1,152 @@
+
+            // backbone
+            backbone: '../../../admin/lib/components-backbone/backbone',
+            backbonerelational: '../../../admin/lib/backbone-relational/backbone-relational',
```
Remove the things you aren't using.

clockard replied:
Removed unused libraries.

shaundmorris commented:
👍 for now, would like to be able to test it with dependent PRs. Looks like there is only one outstanding question.

harrison-tarr commented on catalog/spatial/registry/registry-admin-modules/registry-admin-local-ui/pom.xml:
```diff
@@ -0,0 +1,130 @@
+                                        <limit>
+                                            <counter>INSTRUCTION</counter>
+                                            <value>COVEREDRATIO</value>
+                                            <minimum>0.0</minimum>
```
this seems low, is it just UI code in here or will you be writing tests?

clockard replied:
It is all ui stuff. The one java class is for the ui plugin that doesn't have any logic in it

harrison-tarr commented:
Somethings I've noticed since playing around with the UI: 
when editing a node, the security attributes `?` doesn't seem to have scroll over text. 
The associations dropdown is empty, and the `+` next to it doesn't do anything

harrison-tarr commented on catalog/spatial/registry/registry-admin-modules/registry-admin-local-ui/src/main/webapp/js/model/Association.js:
```diff
@@ -0,0 +1,250 @@
+        },
+        populateFromModel: function (associations) {
+            var model = this;
+            var collectionModel = this.get('associations');
```
I don't know if I've seen it before or if this is a normal thing, but could you replace the strings like 'associations' and 'topSegment' with class constants? I could see one of these being misspelled sometime and causing a huge headache during debugging

clockard replied:
I don't think you normally have constants for strings in js. I haven't seen that in our baseline before. Unless you feel strongly about this I am going to leave this as as.

pklinef replied:
Newer versions of javascript do have constants now but it requires new browsers like IE11 so I would not suggest using it here without using a tool to transpile to es5 which is supported by more browsers.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/const

clockard commented:
@harrison-tarr I have updated the the association section to be disabled unless there are values to choose from. I have also fixed the mouse over issue after you add/remove values from a multi value field

shaundmorris commented:
Is there a convention (like greying out field) for fields that cannot be modified? Like `Last Updated` and `Live Date` fields are disabled and I don't think will ever be modified by an admin.

clockard commented:
Their should probably be something like graying out to indicate they are not editable. I think that could be something address in the usability ticket https://codice.atlassian.net/browse/DDF-2190

tbatie commented on catalog/spatial/registry/registry-admin-modules/registry-admin-local-ui/src/main/webapp/js/model/FieldDescriptors.js:
```diff
@@ -0,0 +1,420 @@
+                Service: {
+                    Name: {
+                        displayName: 'Service Name',
+                        description: 'This service name',
```
service's*

tbatie commented:
+1 Hero'd successfully. Played around with the ui and made sure each new field worked correctly.

pklinef commented on catalog/spatial/registry/registry-admin-modules/registry-admin-local-ui/src/main/webapp/js/model/Association.js:
```diff
@@ -0,0 +1,250 @@
+                }
+            }
+        ],
+        defaults: {
```
Defaults should use a function that returns these values.  As it is, all models will have shared references if anything was default.

clockard replied:
I see. In this case there should only ever be one instance of an AssociationModel. Should this still be changed?

clockard replied:
Changed in DDF-2190

pklinef commented on catalog/spatial/registry/registry-admin-modules/registry-admin-local-ui/src/main/webapp/js/model/Association.js:
```diff
@@ -0,0 +1,250 @@
+                   model.removeSegment(curSeg);
+               }
+            });
+            wreqr.vent.trigger('associationSegmentRemoved',segId);
```
Instead of doing a wreqr event, the views could listen to the segment collection for remove events.  This comments applies to other uses of wreqr in this model.

pklinef commented on catalog/spatial/registry/registry-admin-modules/registry-admin-local-ui/src/main/webapp/js/model/Segment.js:
```diff
@@ -0,0 +1,464 @@
+                }
+            }
+        ],
+        defaults: {
```
Use a function for defaults.

>Remember that in JavaScript, objects are passed by reference, so if you include an object as a default value, it will be shared among all instances. Instead, define defaults as a function.
http://backbonejs.org/#Model-defaults

clockard replied:
I can see how this could be a problem. I always set those fields so it isn't a functional problem right now. If it is ok with you I will make this update in https://codice.atlassian.net/browse/DDF-2190

clockard replied:
Changed in DDF-2190

pklinef commented on catalog/spatial/registry/registry-admin-modules/registry-admin-local-ui/src/main/webapp/js/view/Association.view.js:
```diff
@@ -0,0 +1,90 @@
+                if (selectedOption.length === 1) {
+                    var addedItem = this.model.addAssociation(this.parentId, selectedOption.attr('name'), 'AssociatedWith');
+                    this.collection.add(addedItem);
+                    this.render();
```
Instead of doing the render here, you can add to events that you want to render when the collection changes.  That way the view will update even if something else adds to that collection.

clockard replied:
Changed in DDF-2190

pklinef commented on catalog/spatial/registry/registry-admin-modules/registry-admin-local-ui/src/main/webapp/js/view/Registry.view.js:
```diff
@@ -0,0 +1,186 @@
+                new RegistryView.ModalController({
+                    application: this
+                });
+                this.listenTo(this.model, 'add', this.render);
```
Marionette gives model and view event hashes similar to the DOM events hash.

http://marionettejs.com/docs/v2.4.7/marionette.view.html#viewmodelevents-and-viewcollectionevents

clockard replied:
Changed in DDF-2190

clockard commented:
@pklinef I will work off your comments in the follow on ticket  https://codice.atlassian.net/browse/DDF-2190 if that is ok with you.

pklinef commented:
👍 for this PR with my comments being worked off on DDF-2190.
